src/Sarif/Visitors/RebaseUriVisitor.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT        
+// license. See LICENSE file in the project root for full license information. 
+
+using Microsoft.CodeAnalysis.Sarif.Readers;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    /// <summary>
+    /// A class that, given a variable name (e.x. "%SRCROOT%") and a value (e.x. "C:\src\root\"), rebases the URIs in a SARIF log 
+    /// in order to make a SARIF log not depend on absolute paths/be machine independent.
+    /// </summary>
+    public class RebaseUriVisitor : SarifRewritingVisitor
+    {
+        internal const string BaseUriDictionaryName = "originalUriBaseIds";
+        internal const string IncorrectlyFormattedDictionarySuffix = ".Old";
+
+        private string _baseName;
+        private Uri _baseUri;
+
+        private static JsonSerializerSettings _settings;
+
+        internal static JsonSerializerSettings JsonSerializerSettings
+        {
+            get
+            {
+                if (_settings == null)
+                {
+                    _settings = new JsonSerializerSettings();
+                    _settings.ContractResolver = SarifContractResolver.Instance;
+                }
+                return _settings;
+            }
+        }
+        
+        /// <summary>
+        /// Create a RebaseUriVisitor, with a given name for the Base URI and a value for the base URI.
+        /// </summary>
+        public RebaseUriVisitor(string baseName, Uri baseUri)
+        {
+            _baseName = baseName;
+            _baseUri = baseUri;
+            Debug.Assert(_baseUri.IsAbsoluteUri);
+        }
+
+        public override PhysicalLocation VisitPhysicalLocation(PhysicalLocation node)
+        {
+            PhysicalLocation newNode = base.VisitPhysicalLocation(node);
+            
+            if (string.IsNullOrEmpty(newNode.UriBaseId))
+            {
+                if (_baseUri.IsBaseOf(newNode.Uri) && newNode.Uri.IsAbsoluteUri)
+                {
+                    newNode.UriBaseId = _baseName;
+                    newNode.Uri = _baseUri.MakeRelativeUri(node.Uri);
+                }
+            }
+
+            return newNode;
+        }
+
+        public override Run VisitRun(Run node)
+        {
+            Run newRun = base.VisitRun(node);
+
+            if (newRun.Files != null)
+            {
+                FixFiles(newRun);
```

Comment:
>ixFiles(newRun); [](start = 17, length = 16)

why didn't you do this in an override of VisitFiles? Curious

Reply:
VisitFileData (the only option) does not allow us to change the keys as well, and keeping this all together seemed cleaner.